### PR TITLE
Authentication Filter should not return 403

### PR DESCRIPTION
**Pull request by abertnamanya**
## Issue I worked on

see https://issues.openmrs.org/browse/RESTWS-859

## Checklist: I completed these to help reviewers :)

- [ ] My pull request only contains **ONE single commit**
(the number above, next to the 'Commits' tab is 1).

  No? -> [read here](https://wiki.openmrs.org/display/docs/Pull+Request+Tips) on how to squash multiple commits into one

- [ ] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [ ] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [ ] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [ ] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [ ] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

### CODE REVIEW

File: omod-common/src/main/java/org/openmrs/module/webservices/rest/web/filter/AuthorizationFilter.java [hunk 1]
```diff
@@ -37,71 +37,76 @@
...
+                        // This filter never stops execution. If the user failed to
+                        // authenticate, that will be caught later.
+                    }
+                } else {
+                    HttpServletResponse httpResponse = (HttpServletResponse) response;
+                    httpResponse.sendError(HttpServletResponse.SC_BAD_REQUEST, "Basic header does not have a provided credential");
+                }
+            } else {
+                HttpServletResponse httpResponse = (HttpServletResponse) response;
+                httpResponse.sendError(HttpServletResponse.SC_UNAUTHORIZED, "Wrong credentials provided");
```
Comment (ibacher) [inline omod-common/src/main/java/org/openmrs/module/webservices/rest/web/filter/AuthorizationFilter.java:105]: This else block occurs is `Context.isAuthenticated()` returns true. If that's the case, we want to process the request as normal (from the ticket this is step 1).

File: omod-common/src/main/java/org/openmrs/module/webservices/rest/web/filter/AuthorizationFilter.java [hunk 2]
```diff
@@ -37,71 +37,76 @@
...
+                        Context.authenticate(userAndPass[0], userAndPass[1]);
+                        if (log.isDebugEnabled())
+                            log.debug("authenticated " + userAndPass[0]);
+                    } catch (Exception ex) {
+                        // This filter never stops execution. If the user failed to
+                        // authenticate, that will be caught later.
+                    }
+                } else {
+                    HttpServletResponse httpResponse = (HttpServletResponse) response;
+                    httpResponse.sendError(HttpServletResponse.SC_BAD_REQUEST, "Basic header does not have a provided credential");
```
Comment (ibacher) [inline omod-common/src/main/java/org/openmrs/module/webservices/rest/web/filter/AuthorizationFilter.java:101]: In this case, `basicAuth` is `null`, which means that the `Authorization` header is unset. This should return the 401 error indicated in step 2.1.

```suggestion
                    HttpServletResponse httpResponse = (HttpServletResponse) response;
                    httpResonse.setHeader("WWW-Authenticate", "Basic, OpenMRS-Cookie")
                    httpResponse.sendError(HttpServletResponse.SC_UNAUTHORIZED, "Invalid credentials provided");
```

File: omod-common/src/main/java/org/openmrs/module/webservices/rest/web/filter/AuthorizationFilter.java [hunk 3]
```diff
@@ -37,71 +37,76 @@
...
+        if (request instanceof HttpServletRequest) {
+            HttpServletRequest httpRequest = (HttpServletRequest) request;
+            if (httpRequest.getRequestedSessionId() != null && !httpRequest.isRequestedSessionIdValid()) {
+                HttpServletResponse httpResponse = (HttpServletResponse) response;
+                httpResponse.sendError(HttpServletResponse.SC_UNAUTHORIZED, "Session timed out");
+            }
+
+            if (!Context.isAuthenticated()) {
+                String basicAuth = httpRequest.getHeader("Authorization");
+                if (basicAuth != null) {
```
Comment (ibacher) [inline omod-common/src/main/java/org/openmrs/module/webservices/rest/web/filter/AuthorizationFilter.java:86]: There are other `Authorization` headers that might come in that might be processed elsewhere. For instance OAuth2 uses `Authorization: Bearer <opaque token>`. So here we need to check:

```suggestion
                if (basicAuth != null && basicAuth.startsWith("Basic")) {
```

File: omod-common/src/main/java/org/openmrs/module/webservices/rest/web/filter/AuthorizationFilter.java [hunk 4]
```diff
@@ -37,71 +37,76 @@
...
+                    // this is "Basic ${base64encode(username + ":" + password)}"
+                    try {
+                        basicAuth = basicAuth.substring(6); // remove the leading "Basic "
+                        String decoded = new String(Base64.decodeBase64(basicAuth), Charset.forName("UTF-8"));
```
Comment (ibacher) [inline omod-common/src/main/java/org/openmrs/module/webservices/rest/web/filter/AuthorizationFilter.java:90]: Here we need the check for step 2.2 of the flow from the ticket:

```java
if (StringUtils.isBlank(basicAuth) || !basicAuth.contains(":")) {
	HttpServletResponse httpResponse = (HttpServletResponse) response;
	httpResponse.sendError(HttpServletResponse.SC_BAD_REQUEST, "Invalid credentials provided");
	return;
}
```

**Review by ibacher — CHANGES_REQUESTED** [review]
@abertnamanya Thanks for this! First, please make sure that your IDE is configured [according to OpenMRS's standards](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE). In particular, we should be using tabs rather than spaces and don't mess with the license header at all. It needs to be laid out exactly as it was.

File: omod-common/src/main/java/org/openmrs/module/webservices/rest/web/filter/AuthorizationFilter.java (hunk 4 above)
Comment (abertnamanya) [reply omod-common/src/main/java/org/openmrs/module/webservices/rest/web/filter/AuthorizationFilter.java]: @ibacher please check I made changes to the PR

**Comment by HerbertYiga** [comment]
@abertnamanya   have you tested the changes in the running openmrs instance too?

File: omod-common/src/main/java/org/openmrs/module/webservices/rest/web/filter/AuthorizationFilter.java [hunk 5]
```diff
@@ -84,19 +85,31 @@ public void doFilter(ServletRequest request, ServletResponse response, FilterCha
...
-						// This filter never stops execution. If the user failed to
-						// authenticate, that will be caught later.
+					if (basicAuth != null && basicAuth.startsWith("Basic")) {
+						// this is "Basic ${base64encode(username + ":" + password)}"
+						if (StringUtils.isBlank(basicAuth) || !basicAuth.contains(":")) {
+							HttpServletResponse httpResponse = (HttpServletResponse) response;
+							httpResponse.sendError(HttpServletResponse.SC_BAD_REQUEST, "Invalid credentials provided");
+							return;
+						}
+						try {
```
Comment (ibacher) [inline omod-common/src/main/java/org/openmrs/module/webservices/rest/web/filter/AuthorizationFilter.java:95]:
```suggestion
						}
						
						try {
```

File: omod-common/src/main/java/org/openmrs/module/webservices/rest/web/filter/AuthorizationFilter.java [hunk 6]
```diff
@@ -84,19 +85,31 @@ public void doFilter(ServletRequest request, ServletResponse response, FilterCha
...
+							basicAuth = basicAuth.substring(6); // remove the leading "Basic "
+							String decoded = new String(Base64.decodeBase64(basicAuth), Charset.forName("UTF-8"));
+							String[] userAndPass = decoded.split(":");
+							Context.authenticate(userAndPass[0], userAndPass[1]);
+							if (log.isDebugEnabled())
+								log.debug("authenticated " + userAndPass[0]);
```
Comment (ibacher) [inline omod-common/src/main/java/org/openmrs/module/webservices/rest/web/filter/AuthorizationFilter.java:101]:
```suggestion
							log.debug("authenticated {}", userAndPass[0]);
```

File: omod-common/src/main/java/org/openmrs/module/webservices/rest/web/filter/AuthorizationFilter.java [hunk 7]
```diff
@@ -84,19 +85,31 @@ public void doFilter(ServletRequest request, ServletResponse response, FilterCha
...
-						Context.authenticate(userAndPass[0], userAndPass[1]);
-						if (log.isDebugEnabled())
-							log.debug("authenticated " + userAndPass[0]);
-					}
-					catch (Exception ex) {
-						// This filter never stops execution. If the user failed to
-						// authenticate, that will be caught later.
+					if (basicAuth != null && basicAuth.startsWith("Basic")) {
+						// this is "Basic ${base64encode(username + ":" + password)}"
+						if (StringUtils.isBlank(basicAuth) || !basicAuth.contains(":")) {
```
Comment (ibacher) [inline omod-common/src/main/java/org/openmrs/module/webservices/rest/web/filter/AuthorizationFilter.java:90]: Yeah, this block needs to be after we create `decoded` and then:
```suggestion
						if (StringUtils.isBlank(decoded) || !decoded.contains(":")) {
```

File: omod-common/src/main/java/org/openmrs/module/webservices/rest/web/filter/AuthorizationFilter.java [hunk 8]
```diff
@@ -87,22 +87,25 @@ public void doFilter(ServletRequest request, ServletResponse response, FilterCha
...
+							if (StringUtils.isBlank(decoded) || !decoded.contains(":")) {
+								HttpServletResponse httpResponse = (HttpServletResponse) response;
+								httpResponse.sendError(HttpServletResponse.SC_BAD_REQUEST, "Invalid credentials provided");
+								return;
+							}
+							if (StringUtils.isBlank(basicAuth) || !basicAuth.contains(":")) {
+								HttpServletResponse httpResponse = (HttpServletResponse) response;
+								httpResponse.sendError(HttpServletResponse.SC_BAD_REQUEST, "Invalid credentials provided");
+								return;
+							}
```
Comment (ibacher) [inline omod-common/src/main/java/org/openmrs/module/webservices/rest/web/filter/AuthorizationFilter.java:102]: Almost. It should look like this:

```suggestion
							if (StringUtils.isBlank(basicAuth)) {
								HttpServletResponse httpResponse = (HttpServletResponse) response;
								httpResponse.sendError(HttpServletResponse.SC_BAD_REQUEST, "Invalid credentials provided");
								return;
							}
							
							String decoded = new String(Base64.decodeBase64(basicAuth), Charset.forName("UTF-8"));
							if (StringUtils.isBlank(decoded) || !decoded.contains(":")) {
								HttpServletResponse httpResponse = (HttpServletResponse) response;
								httpResponse.sendError(HttpServletResponse.SC_BAD_REQUEST, "Invalid credentials provided");
								return;
							}
```

File: omod-common/src/main/java/org/openmrs/module/webservices/rest/web/filter/AuthorizationFilter.java [hunk 9]
```diff
@@ -87,22 +87,25 @@ public void doFilter(ServletRequest request, ServletResponse response, FilterCha
...
 							String[] userAndPass = decoded.split(":");
 							Context.authenticate(userAndPass[0], userAndPass[1]);
-							if (log.isDebugEnabled())
-								log.debug("authenticated " + userAndPass[0]);
+							log.debug("authenticated {}" + userAndPass[0]);
 						}
 						catch (Exception ex) {
-							// This filter never stops execution. If the user failed to
-							// authenticate, that will be caught later.
+							log.debug("authentication exception "+ex.getMessage());
```
Comment (ibacher) [inline omod-common/src/main/java/org/openmrs/module/webservices/rest/web/filter/AuthorizationFilter.java:108]: Just let the logger handle the exception:

```suggestion
							log.debug("authentication exception ", ex);
```

File: omod-common/src/main/java/org/openmrs/module/webservices/rest/web/filter/AuthorizationFilter.java [hunk 10]
```diff
@@ -87,22 +87,25 @@ public void doFilter(ServletRequest request, ServletResponse response, FilterCha
...
+							if (StringUtils.isBlank(basicAuth) || !basicAuth.contains(":")) {
+								HttpServletResponse httpResponse = (HttpServletResponse) response;
+								httpResponse.sendError(HttpServletResponse.SC_BAD_REQUEST, "Invalid credentials provided");
+								return;
+							}
 							String[] userAndPass = decoded.split(":");
 							Context.authenticate(userAndPass[0], userAndPass[1]);
-							if (log.isDebugEnabled())
-								log.debug("authenticated " + userAndPass[0]);
+							log.debug("authenticated {}" + userAndPass[0]);
```
Comment (ibacher) [inline omod-common/src/main/java/org/openmrs/module/webservices/rest/web/filter/AuthorizationFilter.java:105]: This produces the log message, e.g., "authenticated {}admin", it should be:

```suggestion
							log.debug("authenticated {}", userAndPass[0]);
```

Which produces "authenticated admin" or maybe even better:


```suggestion
							log.debug("authenticated [{}]", userAndPass[0]);
```

Which produces "authenticated [admin]"

File: omod-common/src/main/java/org/openmrs/module/webservices/rest/web/filter/AuthorizationFilter.java [hunk 11]
```diff
@@ -87,22 +87,25 @@ public void doFilter(ServletRequest request, ServletResponse response, FilterCha
...
 						}
 						catch (Exception ex) {
-							// This filter never stops execution. If the user failed to
-							// authenticate, that will be caught later.
```
Comment (ibacher) [inline omod-common/src/main/java/org/openmrs/module/webservices/rest/web/filter/AuthorizationFilter.java]: Please restore these comments. They are important to understanding the function of this filter.

**Review by ibacher — CHANGES_REQUESTED** [review]

**Review by ibacher — APPROVED** [review]

**Comment by abertnamanya** [comment]
> @abertnamanya In the future, it's a good idea to include the name of the ticket in the commit message as well as to actually fill out the PR template to make it easier for people to review things.

Thank you @ibacher This was My first PR I think by that time I had not read the Openmrs Github push terms

**Pull request merged**